File: 450.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
    if root != nil {
        if key < root.Val {
            root.Left = deleteNode(root.Left, key)
        }else if key > root.Val {
            root.Right = deleteNode(root.Right, key)
        }else {
            if root.Left != nil && root.Right != nil {
                // found
                root.Val = getMin(root.Right).Val
                root.Right = deleteNode(root.Right, root.Val)
            }else if root.Left == nil && root.Right == nil {
                root = nil
            }else if root.Left == nil {
                root = root.Right
            }else {
                root = root.Left
            }
        }
    }
    return root
}

func getMin(this *TreeNode) *TreeNode {
    if this == nil {
        return nil
    }
    for {
        if this.Left == nil {
            return this
        }
        this = this.Left
    }
}

File: tree/450.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root)
            return root;
        if (root->val < key) {
            root->right = deleteNode(root->right, key);
        } else if (root->val > key) {
            root->left = deleteNode(root->left, key);
        } else {
            if (!root->left && !root->right) {
                delete(root);
                return nullptr;
            } else if (!root->left && root->right) {
                auto ret = root->right;
                delete(root);
                return ret;
            } else if (root->left && !root->right) {
                auto ret = root->left;
                delete(root);
                return ret;
            } else {
                auto removal = root->right;
                while (removal->left) {
                    removal = removal->left;
                }
                root->val = removal->val;
                root->right = deleteNode(root->right, root->val);
            }
        }
        return root;
    }
};

